Mchirp and q from uniform mass1 and mass2

yi-fan-wang opened the pull request:
Adding a new distribution for mchirp and q (mass ratio) from uniform mass1 and mass2 (component mass). In addition, there seems to be a bug for "lognorm" in the power law distribution. I have fixed it.

ahnitz commented:
@yi-fan-wang Fix the cc errors. It would be good to see a script that showt that this works though. it should be easy to do. Can you put together a few lines where you draw numbers uniformly from 0->1 and then transform using the invcdf function. You should be able to plot the density of points on the m1-m2 plane and show that it is a uniform distribution.

ahnitz commented:
@yi-fan-wang Can you please post that plot and script to make it?

yi-fan-wang commented:
@ahnitz @cdcapano I have tested the new distribution with this script: https://github.com/yi-fan-wang/GWscripts/blob/master/pycbc/mchirp_q_from_m1_m2.ipynb 
See the figure 'out[26]', it is a uniform distribution in the mass1-mass2 coordinates, but in a triangle-like shape, not square. So this is uniform in mass1 and mass2 but + constraints due to upper and lower limits from mchirp and q

cdcapano commented on pycbc/distributions/mass.py:
```diff
@@ -0,0 +1,194 @@
+
+class MchirpfromUniformMass1Mass2(power_law.UniformPowerLaw):
+    """ For a uniform distribution in volume using spherical coordinates, this
+    is the distriubtion to use for the radius.
```
Update the doc string here. It looks like you just copied from the uniform power law doc string. Just add a description of what this distribution is for.

cdcapano commented on pycbc/distributions/mass.py:
```diff
@@ -0,0 +1,194 @@
+
+class QfromUniformMass1Mass2(bounded.BoundedDist):
+    """
+    A distribution for mass ratio (i.e., q) from uniform component mass. The parameters are
```
Put the first sentence on the same line as the `"""`. Then put a blank line to separate that from the rest.

cdcapano reviewed (changes requested):
Looks good. Just add this to the unit tests, and fix the doc strings (see my comments below).

cdcapano commented on pycbc/distributions/mass.py:
```diff
@@ -0,0 +1,238 @@
+
+class MchirpfromUniformMass1Mass2(power_law.UniformPowerLaw):
+    """ 
+    A distribution for chirp mass from uniform component mass + constraints given by chirp mass.
```
This should start on the same line as the `"""`; e.g., `"""A distribution for...`. Also, limit the line width to 79 characters.

cdcapano commented on pycbc/distributions/mass.py:
```diff
@@ -0,0 +1,238 @@
+
+class QfromUniformMass1Mass2(bounded.BoundedDist):
+    """
+    A distribution for mass ratio (i.e., q) from uniform component mass + constraints given by q.
```
Same as above; start this on the same line as the `"""`.

cdcapano reviewed (changes requested):
Getting there... just address the code climate issues. You can ignore the ones about similar code, python 3, not conforming to snake case, and missing docstring for inherited functions.

yi-fan-wang commented:
@ahnitz @cdcapano  I have addressed all comments (mainly code climate). The search test failed but I don't think it's relevant to this PR. In addition, could you confirm [this](https://github.com/gwastro/pycbc/pull/3681/files#diff-c3d0c52dd6a752c56eb56a5fdd342dd625bb2b44d15fe08a5c75a6c72d4ff1b2R137) is indeed a bug? Otherwise all looks good. I also added some documentation as shown [here](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/_gh-pages/latest/html/distributions.html#generating-mchirp-and-q-from-uniform-mass1-and-mass2) .

cdcapano commented on pycbc/distributions/power_law.py:
```diff
@@ -134,7 +134,7 @@ def __init__(self, dim=None, **params):
                                    (self._bounds[p][1]**(self.dim) -
                                     self._bounds[p][0]**(self.dim))
-            self._lognorm = numpy.log(self._norm)
+            self._lognorm += numpy.log(self._norm)
```
Oh, I missed this. No, undo this. You'll be overcounting the norm for multiple parameters here.

Actually, as I look at it, if you want to improve the code, take the `self._lognorm` out of the loop. There's no reason to recompute the lognorm for every parameter. So, the block of code should be:
```
        self._norm = 1.0
        for p in self._params:
            self._norm *= self.dim  / \
                                   (self._bounds[p][1]**(self.dim) -
                                    self._bounds[p][0]**(self.dim))
        self._lognorm = numpy.log(self._norm)
```

yi-fan-wang replied:
This is fixed and improved according to your suggestion in the [latest commit ](https://github.com/gwastro/pycbc/pull/3681/commits/275a1818197b48367ece03c881b089668dbc1f2d).

cdcapano reviewed (changes requested):
Thanks for all the changes Yifan, especially for adding the documentation! 🥇 

Please fix the norm that you asked about (see my comment). Otherwise, looks good. I'm not sure why the run job failed, I've restarted it. Although it'll restart anyway once you push the change to the norm.

yi-fan-wang commented:
@ahnitz @cdcapano The latest documentation is [here](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/_gh-pages/latest/html/distributions.html#generating-mchirp-and-q-from-uniform-mass1-and-mass2). All seems good to me. Except that the tests still failed.

cdcapano reviewed (approved):
Looks good @yi-fan-wang . I'm not sure why the tests are failing. @ahnitz does Yifan need to rebase?

ahnitz commented:
@yi-fan-wang Please rebase

yi-fan-wang commented:
I made some tests for PE with GW150914. The run with uniform mass1 and mass2 prior is [here](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/testnewdist/m1m2/) (check the [prior.ini](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/testnewdist/m1m2/prior.ini)). The same configuration but with mchirp_from_uniform_m1_m2 and q_from_uniform_m1_m2 is [here](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/testnewdist/mchirprun/)(check the [prior.ini](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/testnewdist/mchirprun/prior.ini)). The result plot are [here](https://www.atlas.aei.uni-hannover.de/work/yifan.wang/pycbcdoc/testnewdist/comparison2.png). They match well!

Here is an example to use this prior in pycbc_inference:
```
[variable_params]
srcmchirp = 
q = 

[prior-srcmchirp]
name = mchirp_from_uniform_mass1_mass2
min-srcmchirp = 10
max-srcmchirp = 50

[prior-q]
name = q_from_uniform_mass1_mass2
min-q = 1
max-q = 4

[waveform_transforms-redshift]
name = custom
inputs = comoving_volume
redshift = redshift_from_comoving_volume(comoving_volume)

[waveform_transforms-mass1]
name = custom
inputs = srcmchirp, q, redshift
mass1 = mass1_from_mchirp_q(srcmchirp,q) * (1 + redshift)

[waveform_transforms-mass2]
name = custom
inputs = srcmchirp, q, redshift
mass2 = mass2_from_mchirp_q(srcmchirp,q) * (1 + redshift)
```

cdcapano commented:
Great! Thanks @yi-fan-wang .
